Sources/GRPCCore/GRPCServer.swift

```diff
@@ -204,140 +204,42 @@ public struct GRPCServer: Sendable {
+      for transport in self.transports {
         group.addTask {
-          await withDiscardingTaskGroup { subGroup in
-            do {
-              for try await stream in listener {
-                subGroup.addTask {
-                  await self.router.handle(stream: stream, interceptors: self.interceptors)
-                }
-              }
-            } catch {
-              // If the listener threw then the connection must be broken, cancel all work.
-              subGroup.cancelAll()
+          do {
+            try await transport.listen { stream in
+              await self.router.handle(stream: stream, interceptors: self.interceptors)
             }
+          } catch {
+            throw RuntimeError(
+              code: .failedToStartTransport,
+              message: """
+                Server didn't start because the '\(type(of: transport))' transport threw an error \
+                while starting.
+                """,
+              cause: error
+            )
           }
         }
       }
-    }
-  }
-  #endif
 
-  private func handleRequestsInTaskGroup(_ listeners: [RPCAsyncSequence<Stream>]) async {
-    // If the discarding task group isn't available then fall back to using a regular task group
-    // with a limit on subtasks. Most servers will use an HTTP/2 based transport, most
-    // implementations limit connections to 100 concurrent streams. A limit of 4096 gives the server
-    // scope to handle nearly 41 completely saturated connections.
-    let maxConcurrentSubTasks = 4096
-    let tasks = ManagedAtomic(0)
+      self.state.store(.running, ordering: .sequentiallyConsistent)
 
-    await withTaskGroup(of: Void.self) { group in
-      for listener in listeners {
-        group.addTask {
-          await withTaskGroup(of: Void.self) { subGroup in
-            do {
-              for try await stream in listener {
-                let taskCount = tasks.wrappingIncrementThenLoad(ordering: .sequentiallyConsistent)
-                if taskCount >= maxConcurrentSubTasks {
-                  _ = await subGroup.next()
-                  tasks.wrappingDecrement(ordering: .sequentiallyConsistent)
-                }
+      while !group.isEmpty {
+        do {
+          try await group.next()
+        } catch {
+          // Failed to start some transport, so start stopping.
+          self.state.store(.stopping, ordering: .sequentiallyConsistent)
 
-                subGroup.addTask {
-                  await self.router.handle(stream: stream, interceptors: self.interceptors)
-                }
-              }
-            } catch {
-              // If the listener threw then the connection must be broken, cancel all work.
-              subGroup.cancelAll()
-            }
+          // Some transports may have started and have streams which need closing:
+          // tell them to stop listening.
+          for transport in self.transports {
+            transport.stopListening()
           }
```

Comment:
Is stopping the right thing or should we just cancel? It could be really problematic if somebody starts a server with multiple transports and then one fails but the service doesn't exit because the graceful shutdown is still ongoing. IMO a hard cancellation here is better.

Reply:
I am doubtful that this can ever be the case. If a transport dies and there is no way it self-heals your service seems pretty doomed. We can make this configurable but I would suggest we start with a forceful cancellation and then in the future can add a config property `shouldGracefullyShutdownOnTransportFailure`. (We are getting pretty close to re-invent service lifecycle with that one though ;) )

Reply:
Sure. My reasoning is that if a single transport fails it is often detrimental to the health of the service if this results in us gracefully shutting down the whole service we have a potentially long period where our service is in a degraded state. That's why I would argue that if a transport fail it should lead to us closing the whole server down as quickly as possible so that it can potentially restart and recover.

I can see arguments where people would want to gracefully shutdown but in my personal opinion this should not be the default.

Reply:
Thanks for elaborating. I understand your point but I'm not yet convinced that cancellation is the right behaviour and I don't really like the idea of having configuration for the behaviour either (as it introduces more edges to test).

I've not really followed the thinking through yet, but I wonder if the server should have some kind of async sequence for its state. That way it could report, for example, when all transports have started, and when a transport throws (i.e. in some degraded state). This gives the caller control over what to do (e.g. let it shutdown gracefully, or cancel). Suspect we'd need to think a little bit more about what state/events could happen though.